### ISSUE TITLE
Workaround for non-scrollable immediate parent?

### ISSUE THREAD

**Issue by lordakshaya** [brigade/react-waypoint#55]
Hey @trotzig,

Is there a work around for the following limitation?

>We also base the visibility calculations on the scroll position of the scrollable container (or window if no scrollable container is found). This means that if your scrollable container has a height that is greater than the window, it might trigger onEnter unexpectedly.

Currently, I have a DOM structure like:

```
Body
  --> Header
  --> Content
            --> Photos                 
                   --> Photo 1
                   --> Photo 2
                   --> Photo 3
                        ....
                   --> Photo N
```

I use Waypoint in the Photo component and I am using the onEnter event to keep track of how many times a Photo has been seen by the user (and hence the `fireOnRapidScrolling={false}` flag). The problem is that none of my div containers have the overflow property and due to which the *onEnter* event is fired for all the Photo components as soon as they are added to the DOM even if they aren't visible in the viewport. 

Any idea on how best to take care of this problem? 

P.S. I'd like to have my page to have the infinite scroll instead of the "Photos" div.

**UPDATE**
I added the following CSS props
```css
.content {
   overflow-y: auto; 
   position: relative;
}
```
but it didnt help either.

**UPDATE 2**
Removed the above CSS and instead added `{threshold: -0.2}`, the behaviour is closer to what I expect but its still not pixel perfect. And this threshold value I arrived is more via some hit and trial, any robust way to figure this out?

**Comment by trotzig** [comment]
Do the images have a fixed height when they are loaded (i.e. `<img height="200">` or `<img style="height: 200px">`)? If not, the waypoints will all be at the top of the page on initial render/load. 

Unless you have a `overflow: auto|scroll` and a `height: <some big number>px` you shouldn't be affected by the limitation above.

**Comment by lordakshaya** [comment]
@trotzig 
1. I dont have the overflow property set anywhere (setting it didnt help either, see the update on the original post)
2. The Photo div has a `min-height` set to 100px, does that count?

**Comment by lordakshaya** [comment]
@trotzig I found the cause of the strange behaviour. I added a breakpoint at the handleScroll method looked at the call stack. I observed that the initial set of calls to _handleScroll were going thanks to the *componentDidMount* method, now the problem is...when this initial set of calls are made, at that moment the CSS styles haven't applied. So, the elements are taking the minimal height they should and hence more of these elements are visible on the screen, e.g.

Lets say the **Photos** container is scrollable and has a natural height of 100px (not set thru CSS), and each **Photo** component is set to have a min-height of 50px **BUT** if the CSS isn't applied, lets says each Photo component would take only 25px. The desired behaviour is, at the time of pageLoad the onEnter event should be fired for first 2 Photos only but since the CSS isn't applied correctly, the onEnter event is fired for first 4 photos!! (didn't check for onLeave event). After the load, the onEnter for the 5th photo component is fired properly i.e. when the waypoint is visible on the viewport.

What do you think can be done here?
if height of the scrollable ancestor is say 100px and that there 5 elements in this vertical container.

**Comment by lencioni** [comment]
Do you know why the CSS isn't applied at that time?

**Comment by lordakshaya** [comment]
@lencioni Prolly because of the design of my Photo Component, the code looks like:

```jsx
render() {
 <div className="photoComp">
  <img {...props} />
  <Waypoint onEnter={} />
 </div>
}
```

What I am thinking is Waypoint's componentDidMount method is called before the Photo Component's render method finishes applying CSS and hence the cause of the problem. WDYT?

**Comment by trotzig** [comment]
I don't think this is because the css is applied after `componentDidMount`, unless you have some elaborate lazy-loading setup for your css. 

One thing to note is that waypoint uses a `<span>` to render. Since `<img>` is rendered with `display: inline` by default, the waypoint is probably right next to the image (at least before it loads), not under it. You could try wrapping either the `<img>` or `<Waypoint>` in a `<div>` and see what happens. I would also consider setting a static width and height (or min-width/min-height) on the `<img>`. I guess you might be doing that already (the `...props` might include it), but just to be sure.

**Comment by lordakshaya** [comment]
@trotzig dude...you hit it right at the spot. So, my CSS was all good that wasnt the problem and to prove my point I even replaced the images with a random div of fixed height and was still facing the issue. However, the point which did make more sense to me was the *lazy-loading* setup for CSS. So, I am using the https://github.com/erikras/react-redux-universal-hot-example boilerplate and what I did realise sometime back was that in the developer mode, in order to add the redux dev-tools, it uses a lazy-loading mechanism! Once, I disabled the redux debug mode or ran the app in production mode...react-waypoint started acting up sweetly and smartly :dancer:  Closing the issue now! Hopefully, some day another developer who may get stuck with redux dev tools and waypoint issue might learn something from this! Thanks @lencioni for taking a look as well.
